tests/FsCodec.SystemTextJson.Tests/UnionConverterTests.fs

```diff
@@ -0,0 +1,361 @@
+
+open Samples
+
+// TODO support [<Struct>]
+type TestRecordPayload =
+    {
+        test: string
+    }
+
+// TODO support [<Struct>]
+type TrickyRecordPayload =
+    {
+        Item: string
+    }
+
+[<JsonConverter(typeof<TypeSafeEnumConverter<Mode>>)>]
+type Mode =
+    | Fast
+    | Slow
+
+[<NoComparison>] // NB this is not a general restriction; it's forced by use of Nullable<T> in some of the cases in this specific one
+[<JsonConverter(typeof<UnionConverter<TestDU>>)>]
+type TestDU =
+    | CaseA of TestRecordPayload
+    | CaseB
+    | CaseC of string
+    | CaseD of a: string
+    | CaseE of string * int
+    | CaseF of a: string * b: int
+    | CaseG of TrickyRecordPayload
+    | CaseH of a: TestRecordPayload
+    | CaseI of a: TestRecordPayload * b: string
+    | CaseJ of a: Nullable<int>
+    | CaseK of a: int * b: Nullable<int>
+    | CaseL of a: Nullable<int> * b: Nullable<int>
+    | CaseM of a: int option
+    | CaseN of a: int * b: int option
+    | CaseO of a: int option * b: int option
+    | CaseP of CartId
+    | CaseQ of SkuId
+    | CaseR of a: CartId
+    | CaseS of a: SkuId
+    | CaseT of a: SkuId option * b: CartId
+    | CaseU of SkuId[]
+    | CaseV of skus: SkuId[]
+    | CaseW of CartId * SkuId[]
+    | CaseX of a: CartId * skus: SkuId[]
+    | CaseY of a: Mode * b: Mode
+    | CaseZ of a: Mode * b: Mode option
+
+#nowarn "1182" // From hereon in, we may have some 'unused' privates (the tests)
+
+// no camel case, because I want to test "Item" as a record property
+// Centred on ignoreNulls for backcompat; round-tripping test covers the case where they get rendered too
+let defaultOptions = Options.Create(converters = [| UnionConverter<TestDU> () |], camelCase = false, ignoreNulls = true)
+
+
+[<Fact>]
+let ``produces expected output`` () =
+    let serialize (x : obj) = JsonSerializer.Serialize(x, defaultOptions)
+    let a = CaseA {test = "hi"}
+    test <@ """{"case":"CaseA","test":"hi"}""" = serialize a @>
+
+    let b = CaseB
+    test <@ """{"case":"CaseB"}""" = serialize b @>
+
+    let c = CaseC "hi"
+    test <@ """{"case":"CaseC","Item":"hi"}""" = serialize c @>
+
+    let d = CaseD "hi"
+    test <@ """{"case":"CaseD","a":"hi"}""" = serialize d @>
+
+    let e = CaseE ("hi", 0)
+    test <@ """{"case":"CaseE","Item1":"hi","Item2":0}""" = serialize e @>
+
+    let f = CaseF ("hi", 0)
+    test <@ """{"case":"CaseF","a":"hi","b":0}""" = serialize f @>
+
+    let g = CaseG {Item = "hi"}
+    test <@ """{"case":"CaseG","Item":"hi"}""" = serialize g @>
+
+    // this may not be expected, but I don't intend changing it
+    let h = CaseH {test = "hi"}
+    test <@ """{"case":"CaseH","test":"hi"}""" = serialize h @>
+
+    let i = CaseI ({test = "hi"}, "bye")
+    test <@ """{"case":"CaseI","a":{"test":"hi"},"b":"bye"}""" = serialize i @>
+
+    let p = CaseP (CartId.Parse "0000000000000000948d503fcfc20f17")
+    test <@ """{"case":"CaseP","Item":"0000000000000000948d503fcfc20f17"}""" = serialize p @>
```

Comment:
TODO: Test fails. 
```json
"{\"case\":\"CaseP\",\"Item\":\"0000000000000000948d503fcfc20f17\"}" = "{\"case\":\"CaseP\",\"Value\":\"0000000000000000948d503fcfc20f17\"}"
false
```
Likely because [`IgnoreDataMember` isn't a thing in STJ](https://github.com/dotnet/runtime/issues/30505). [Replacement is `JsonIgnore`](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonignoreattribute?view=netcore-3.0) but that means updating [the type](https://github.com/jet/FsCodec/pull/43/files#diff-981875b989afa337442baac193423291R46) so switching to STJ wouldn't just be a matter of changing libraries.

Reply:
Nope, not `IgnoreDataMember`. Turns out it's the converter not being picked (in this case the CartId pickler), same [as mentioned above](https://github.com/jet/FsCodec/pull/43/files#r400596805).